### ISSUE TITLE
linux semaphore adds overhead in bulk operations for the blocking version

### ISSUE THREAD

**Issue by samanbarghi** [cameron314/concurrentqueue#36]
I started using the blocking version of the concurrent queue for scheduling tasks in Linux. The performance is pretty good, however I noticed that using bulk operations there is a bottleneck due to calling `semaphore_signal` in a loop. The bulk operation gets rid of the atomic operation overhead but since the semaphore_signal is being called multiple times, and being a system call it's an expensive operation, the performance starts to degrade after there are more than 15 consumer threads where there are 20 tasks per consumers. So basically this part if problematic: 

https://github.com/cameron314/concurrentqueue/blob/master/blockingconcurrentqueue.h#L127

However, I am not sure if there is any alternatives for this since linux does not provide a low over head alternative. So just wanted to point it out , you might have an idea on how to remove this bottleneck?

**Comment by cameron314** [comment]
Hi! Thanks for opening an issue. I have lots of questions (sorry!):

First, how are you determining where the time is going? Are you sure it's mostly semaphore syscall overhead?

Second, that's actually the semaphore code for Mac OS X/iOS. You probably mean [this method](https://github.com/cameron314/concurrentqueue/blob/master/blockingconcurrentqueue.h#L175) instead? It's implemented the same way, though (with a loop).

When you say there's 20 tasks per consumer, does that mean ~20 tasks dequeued with each call to `wait_dequeue_bulk`? Or 20 tasks spread out over a period of time? When the consumer dequeues, does that often empty the queue? Or does it leave it partially full most of the time?

How many producers do you have (roughly)? Also, are you using consumer tokens?

There generally shouldn't be much semaphore overhead, since the system-level semaphore is only signaled if somebody is waiting on it (and similarly only waited on if the queue is empty). The queue would have to be constantly completely emptied, waited on by many consumers, then refilled cyclically to trigger lots of syscalls. If this is what's happening, that means you probably have too many consumers -- they suck the queue dry then all sit still waiting for elements. And if they're waiting anyway, surely a bit of syscall overhead is acceptable? Because they'd be sitting there in a busy loop otherwise no better off... (the semaphore wrapper *does* spin-wait a bit first, by the way).

Note that the number of semaphore signals done is always less than or equal to the number of waiting consumers at the time the elements are enqueued. The maximum syscall overhead is only two syscalls per consumer: one to start waiting when the queue is empty, and one to tell a waiting consumer that there's something to consume. The number of elements actually passed through the queue (and bulk vs single, etc.) is orthogonal to this overhead, which has to do purely with wait-state changes.

**Comment by samanbarghi** [comment]
Hi! I'll be happy to answer them.  

I am using zoom (http://www.rotateright.com/) to profile the application and from that I can see a high amount of time spent on sem_post. The performance degradation can be due to the single poller thread (explained later) becoming the bottleneck, but I can still see the overhead of sem_post have some effect. It's worth mentioning that the results are good and the problem occurs under very heavy workloads. 

Oh, and yes I meant the linux implementation (for some reason I searched for semaphore_signal and posted that! hmm, holidays had it's effect on me!)

So let me explain the scenario a bit further so it becomes more clear. The experiment I am running consist of _n_ worker threads (kernel threads) and there are _20*n_ user threads that each contains a connection. Also, there is a poller kernel thread which runs _epoll\_wait_ periodically. I have two queues! _worker\_queue_ and _poller\_queue_, each relating to worker threads and the poller thread. 

The worker threads dequeue users threads (connections) from the _worker\_queue_, and if there is a read/write (send/recv) on the network and the socket is not ready they enqueue that to the _poller\_queue_, and dequeue another one from the _worker\_queue_. Poller thread just dequeue from _poller\_queue_ add to epoll and when ready enqueue the user thread back to on the _worker\_queue_. Also poller thread bulk enqueue all the results that are back from *epoll_wait*.

Now to answer your questions: all threads are using consumer tokens and bulk operations. Threads dequeue _num\_user\_threads\_in\_queue/num\_threads_ with each *wait_dequeue_bulk* for fairness. So if there is more than one task (user thread) on the queue dequeue operation does not always empty the queue (unless it's the last chunk). 

Sorry the _20_ is a bit confusing, so just assume there are _20*n_ total tasks in the system, and depending on the situation they can be in the global or local queue or be running. 

But I guess you are right. The case is that the operation each worker thread does is very small and the traffic on the poller thread increases. It can be possible that many worker threads are waiting on an empty *worker_queue* while the poller is waiting for connections to be ready to read from or write to (the client in this case is aggressively sending request and waiting for responses from the server that I don't expect the poller thread need to wait for connections for a long time). The syscall overhad is acceptable for sure, but I assume in this case it could help if one could do the sem_post in bulk too (although Linux does not provide any tools for that, an idea like Flexsc would come in handy though). 

And yes, the waiting overhead is a necessity and can not be removed, but I guess my scenario would have a higher performance in windows than in Linux as you can increment the semaphore in bulk. (well, at least in theory! ). But as I said it probably is not a big deal for general use cases as the queue performs very well, only in this specific case with many consumers and heavy workload this can cause a bit of problem.

**Comment by cameron314** [comment]
Zoom looks pretty cool. I'll have to try it next time I want to profile something on Linux (yay for remote Windows GUI support!).

Alright, I have a much better idea of how you're using the queue, thanks.

I agree it'd be quite nice to have bulk semaphore signalling, but I don't know of any way either (though I'm not a Linux expert, mind).

I'm afraid I can't offer much in the way of help -- it seems that everything is working the way it was designed to. Perhaps you could try lowering the number of worker threads (increasing the factor from 20 to say 40?). It would certainly affect the number of waits done, though whether it would result in better or worse total application performance I can't say (it's even possible that total throughput increases but at the cost of additional latency). You could also play with the number of items bulk-dequeued.

One unfortunate factor is that given, say, ten waiting consumers, an enqueue operation of 15 items doesn't know that those consumers are going to do bulk consumptions, so it signals the semaphore 15 times waking up all ten consumers, which each get one item (and fight over the leftover 5). That means that regardless of the fact that you're using bulk operations, when there's waiting consumers, the queue always prefers to wake up as many as it can to waking up a single one (this actually makes sense and is desired in general, it's just a little unfortunate in this case with low-overhead processing on the dequeued items combined with bulk dequeue and many consumers).

I'm going to close the issue, but if you have any more comments or questions please add them here (or if you discover a bug feel free to reopen).
